free-3.md

```diff
@@ -0,0 +1,230 @@
+# Free project 2++
+
+[I see **trees** of green](https://www.youtube.com/watch?v=A3yCcXgbKrE)
+
+## The user and a language
+This section describes who the project would serve and why a language might be a
+good way to meet their needs.
+
+
+### What's the need?
+
+The users I have in mind for this DSL are programmers who are interested in
+rapidly writing binary tree algorithms. They want to be able to test out ideas
+quickly and elegantly, and may not care very much about systems-level
+optimizations.
+
+Right now they people probably write a lot of boilerplate code for each tree
+they write, including the structure of nodes, code for tree rotations,
+debugging tools and more.
+
+Furthermore the details of some of the boilerplate may vary between trees
+written by a single author and definitely vary between trees written by
+different authors, making it just a bit harder to understand a tree written by
+someone else.
+
+I think that standardizing the way in which trees are written would help users
+understand what each other have written and would also save a lot of time spent
+writing boilerplate. This would also help users experiment with different ideas
+for tree algorithms more quickly.
+
+### Why a language?
+
+I think that this problem could be solved with an internal DSL.
+
+The DSL could provide a standardized way to specify the information being
+stored in the tree and the way algorithms which manipulate the structure of the
+tree or perform computations on it should work. Language seems like the natural
+way to do this standardization.
+
+A DSL would also have the ability to take most of the boilerplate out of the
+hands of the users, much in the say way that ScalaTest dramatically reduces the
+amount of code that needs to be written to do unit testing.
+
+One might reasonably suggest that creating some sort of abstract class which
+can be subclassed would be sufficient to solve this problem. The core of this
+observation is the question of whether such an implementation is really
+creating an internal DSL. I claim that it is. The result of people using such a
+super-class would be people writing different trees that all use a common
+vocabulary inherited from the super-class. I don't think this is the only way
+to solve the problem, but even if this was the solution I think it could be
+considered a DSL.
+
+### Why you?
+
+I find binary trees to be exceptionally interesting (dare I say magical?)
+data structures, and have spent a reasonable amount of time writing different
+types of trees in different ways. I think it would be really cool to have a
+standardized way of writing tree algorithms that gives you a whole bunch of
+debugging information and analysis tools right off the bat.
```

Comment:
Debuggers/ IDE's are designed to be used in many situations. Visually presenting tree-structured data isn't their strong point (probably because they all operate on the single-object / single-memory location model).